### PR TITLE
Data.Dict: Add .lookup()

### PR DESCRIPTION
**Pull request by aiya000**
Haskell like .lookup() 😄

### CODE REVIEW

File: doc/vital/Data/Dict.txt [hunk 1]
```diff
@@ -113,5 +113,17 @@ foldr({f}, {init}, {dict})		*Vital.Data.Dict.foldr()*
 	TODO
 	Behaves like Haskell's Data.Map.Lazy.foldrWithKey().
 
+
+lookup({key}, {dict})			*Vital.Data.Dict.lookup()*
+	If the dict has the key, returns Optional.some({dict}[{key}]).
+	If the dict hasn't the key, returns Optional.none().
```
Comment (ujihisa) [inline doc/vital/Data/Dict.txt]: s/hasn't/doesn't have`

Comment (ujihisa) [inline doc/vital/Data/Dict.txt]: Optionalのdocへのlinkもほしい

**Comment by ujihisa** [comment]
👍 :sushi: looks good except for a minor typo.

It may be also helpful if there's a doc in Optional side to mention about this Data.Dict.lookup.

**Comment by haya14busa** [comment]
I think we should add lookup method to Data.Optional.

I prefer fewer dependencies, especially if the dependent module is rarely used.
Data.Optional is completely different paradigm too. I think almost all Data.Dict users doesn't use lookup method.

Adding dependencies to Data.Dict means all Vim plugins which depends on Data.Dict will include and load Data.Optional too, even if they don't use lookup method.

File: autoload/vital/__vital__/Data/Dict.vim [hunk 2]
```diff
@@ -95,6 +104,12 @@ function! s:foldr(f, init, dict) abort
   return s:_foldl(a:f, a:init, reverse(items(a:dict)))
 endfunction
 
+function! s:lookup(key, dict) abort
```
Comment (haya14busa) [inline autoload/vital/__vital__/Data/Dict.vim]: I think the order of arguments should be dict -> key to be consistent with builtin has_key() function.

I suppose Haskell uses key -> dict order, but it's Vim script after all.

**Comment by aiya000** [comment]
I finished it except namespace locating :+1:

mmm,
How do you think to create Data.Optional.Dict ?

**Comment by rhysd** [comment]
Just a question: Why don't you use `get(dict, key, default)`?

**Comment by rhysd** [comment]
My suggestion: Move this to `s:Optional.lookup(collection, elem)` where `collection` is a list or dict. I think supporting both list and dict would be more useful.

**Comment by aiya000** [comment]
!!

I forgot `get()` :joy:  
`get()` satisfies my use case...

**Comment by aiya000** [comment]
If nobody needs this, I think this should be removed.

If we keep this, its idea is good.

**Comment by ujihisa** [comment]
The potential benefit of returning Some instead of specifying return value by `default` of `get` is chaining, but yeah it's Vim script.

I close this and if somebody want this feature that person can reopen again. Probably not as `Data.Dict.lookup` but as a function in optional side.

**Comment by aiya000** [comment]
@ujihisa Thanks 👍

**Pull request closed without merging**